### ISSUE TITLE
Unknown variable reported with .format but not with f-string?

### ISSUE THREAD

**Issue by Holt59** [microsoft/pyright#5980]
**Describe the bug**

Using `.format()` instead of a format string shows a different behavior about an unknown variable type that does not make sense (to me).

**Code or Screenshots**
The code below shows:

```
Argument type is partially unknown
  Argument corresponds to parameter "args" in function "format"
  Argument type is "str | Unknown" Pylance(reportUnknownArgumentType)
```

...with `.format()`, but not with f-string.

```python
from __future__ import annotations

class C:
    name: str
    outer: C | None

    def f(self) -> str:
        name = self.name
        oc = self.outer
        while oc is not None:
            # name = f"{oc.name}.{name}"
            name = "{}.{}".format(oc.name, name)
            oc = oc.outer

        return name
```

**VS Code extension or command-line**

Both:

- VS Code extension, Pylance v2023.9.12.
- pyright version 1.1.327

**Comment by erictraut** [comment]
Thanks for the bug report. This is due to a bug that occurs when an overloaded function (such as `format`) is called within a loop and the narrowed type of an argument (`name`) depends on the return type of that call. It doesn't affect f-string type evaluations because there are no overloaded calls involved in that case.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.328, which I just published. It will also be included in a future release of pylance.
